plutus-tx/src/PlutusTx/Ratio.hs

```diff
@@ -4,65 +4,318 @@
+{-# INLINEABLE ratio #-}
+ratio :: Integer -> Integer -> P.Maybe Rational
+ratio n d
+  | d P.== P.zero = P.Nothing
+  | d P.< P.zero = P.Just (P.negate n % P.negate d)
+  | P.True =
+    let gcd' = euclid n d
+     in P.Just P..
+        Rational (n `Builtins.quotientInteger` gcd') P.$
+        d `Builtins.quotientInteger` gcd'
+
+-- | Converts a 'Rational' to a GHC 'Ratio.Rational', preserving value. Does not
+-- work on-chain.
+toGHC :: Rational -> Ratio.Rational
+toGHC (Rational n d) = n Ratio.% d
+
+-- | Returns the (possibly reduced) numerator of its argument.
+{-# INLINEABLE numerator #-}
+numerator :: Rational -> Integer
+numerator (Rational n _) = n
+
+-- | Returns the (possibly reduced) denominator of its argument. This will
+-- always be greater than 1, although the type does not describe this.
+{-# INLINEABLE denominator #-}
+denominator :: Rational -> Integer
+denominator (Rational _ d) = d
+
+-- | 0.5
+{-# INLINEABLE half #-}
+half :: Rational
+half = Rational 1 2
+
+-- | Converts an 'Integer' into the equivalent 'Rational'.
+{-# INLINEABLE fromInteger #-}
+fromInteger :: Integer -> Rational
+fromInteger num = Rational num P.one
+
+-- | Converts a GHC 'Ratio.Rational', preserving value. Does not work on-chain.
+fromGHC :: Ratio.Rational -> Rational
+fromGHC r = Ratio.numerator r % Ratio.denominator r
+
+-- | Produces the additive inverse of its argument.
+--
+-- = Note
+--
+-- This is specialized for 'Rational'; use this instead of the generic version
+-- of this function, as it is significantly smaller on-chain.
+{-# INLINEABLE negate #-}
+negate :: Rational -> Rational
+negate (Rational n d) = Rational (P.negate n) d
+
+-- | Returns the absolute value of its argument.
+--
+-- = Note
+--
+-- This is specialized for 'Rational'; use this instead of the generic version
+-- of this function, as it is significantly smaller on-chain.
+{-# INLINEABLE abs #-}
+abs :: Rational -> Rational
+abs rat@(Rational n d)
+  | n P.< P.zero = Rational (P.negate n) d
+  | P.True = rat
+
+-- | @'properFraction' r@ returns the pair @(n, f)@, such that all of the
+-- following hold:
+--
+-- * @'fromInteger' n 'P.+' f = r@;
+-- * @n@ and @f@ both have the same sign as @r@; and
+-- * @'abs' f 'P.<' 'P.one'@.
+{-# INLINEABLE properFraction #-}
+properFraction :: Rational -> (Integer, Rational)
+properFraction (Rational n d) =
+  (n `Builtins.quotientInteger` d,
+   Rational (n `Builtins.remainderInteger` d) d)
+
+-- | Gives the reciprocal of the argument; specifically, for @r 'P./='
+-- 'P.zero'@, @r 'P.*' 'recip' r = 'P.one'@.
+--
+-- = Important note
+--
+-- The reciprocal of zero is mathematically undefined; thus, @'recip' 'P.zero'@
+-- will error. Use with care.
+{-# INLINEABLE recip #-}
+recip :: Rational -> Rational
+recip (Rational n d)
+  | n P.== P.zero = Builtins.error ()
+  | n P.< P.zero = Rational (P.negate d) (P.negate n)
+  | P.True = Rational d n
+
+-- | Returns the whole-number part of its argument, dropping any leftover
```

Comment:
Why not use the same description as `truncate` from the Haskell prelude? (This isn't wrong, just wondering)

Reply:
Because I find this clearer.